# Project Team 35

Team Members:
1. CUI YILAI (A0329782N) :  Formalization of Sudoku Rules into Propositional Form
2. XU CHENHAO (A0332422U) : Answer the Queries 
3. YANG FEIYANG (A0357160H) : Solve the Puzzle
4. YANUARDI FAJRI (A0328886H) : Streamlit Integration

# Sudoku Knowledge Representation & Inference

You will implement three functions -- `build_general_kb`, `build_definite_kb`, `pl_bc_entails` -- and the full-grid solving logic in `sudoku_solver.py`.

This notebook imports and tests those functions. The Streamlit app (`sudoku_app.py`) must import the same implementation from `sudoku_solver.py`; do not copy or rewrite the solver functions inside the app.

**Rules:**

- In `sudoku_solver.py`, import only from `utils.py` and `logic_.py`; do not modify either file.
- Do not duplicate the core solver functions in this notebook or `sudoku_app.py`.
- All other content in this notebook may be edited freely.


In [1]:
from utils import *
from logic_ import *
import json
import time
import importlib
import sudoku_solver

# Reload so edits made to sudoku_solver.py are picked up when this cell is rerun.
importlib.reload(sudoku_solver)

from sudoku_solver import (
    atom,
    build_general_kb,
    build_definite_kb,
    solve_full_grid_fc,
    pl_bc_entails,
    solve_full_grid_bc,
)


## Loading a puzzle from JSON

Puzzles are provided as JSON, not embedded in this notebook. Each file looks like:

```json
{
  "n": 9, "box_h": 3, "box_w": 3,
  "puzzles": [
    {
      "givens": {"1_1": 3, "2_3": 1, ...},
      "given_count": 28,
      "solution": {"1_1": 3, "1_2": 4, ...}
    },
    ...
  ]
}
```

`"r_c"` string keys map to the value at row `r`, column `c` (1-indexed). `given_count` is exactly how many cells are given. `solution` is included so you can check your own work as you go, but your functions must not read `solution` to answer a query, they should only read `givens`.

In [2]:
#do not change this function, it is used to load the puzzle pool from a json file
def load_pool(path):
    with open(path) as f:
        raw = json.load(f)
    puzzles = []
    for p in raw['puzzles']:
        givens = {tuple(int(x) for x in k.split('_')): v for k, v in p['givens'].items()}
        solution = {tuple(int(x) for x in k.split('_')): v for k, v in p['solution'].items()}
        puzzles.append({'givens': givens, 'solution': solution, 'given_count': p['given_count']})
    return raw['n'], raw['box_h'], raw['box_w'], puzzles

n, box_h, box_w, puzzle_pool = load_pool('puzzles.json')
print(f'{len(puzzle_pool)} puzzles loaded, {n}x{n} grid, {box_h}x{box_w} boxes')
print('given_count values:', sorted(p['given_count'] for p in puzzle_pool))

# Pick one puzzle to work with through the rest of this notebook.
puzzle = puzzle_pool[0]
givens = puzzle['givens']
print(f"working puzzle has {puzzle['given_count']} givens")
for r in range(1, n + 1):
    print([givens.get((r, c), '.') for c in range(1, n + 1)])

5 puzzles loaded, 9x9 grid, 3x3 boxes
given_count values: [30, 33, 36, 39, 42]
working puzzle has 30 givens
['.', 3, '.', '.', '.', '.', '.', '.', '.']
[2, '.', '.', '.', '.', 7, 8, 6, '.']
[5, 8, '.', 2, 6, '.', '.', 3, '.']
[7, 5, '.', '.', '.', '.', '.', 8, '.']
['.', '.', '.', '.', 7, '.', 5, '.', 4]
['.', '.', '.', 5, 3, '.', '.', 9, 6]
['.', 1, 2, '.', '.', 9, '.', '.', '.']
[6, 4, '.', '.', 5, 8, 9, '.', '.']
['.', '.', '.', '.', 2, 3, '.', '.', '.']


Notice: `pl_fc_entails`/`pl_resolution`/`tt_entails` are all given to you, already implemented, in `logic_.py`. The one algorithm you write yourself in this assignment is **backward chaining** (`pl_bc_entails`) -- see Task 2.

## Define Symbols

Two families of propositional symbols, for row $r$, column $c$, value $v$ (all ranging over $1$ to $n$):

| Symbol | Meaning |
|---|---|
| $\mathit{Is}_{rcv}$ | cell $(r,c)$ has value $v$ |
| $\mathit{Not}_{rcv}$ | cell $(r,c)$ does **not** have value $v$ |

For each representation in Task 1, determine which of these two families is actually needed.

Hint: consider what form a definite (Horn) clause must take, and what `PropDefiniteKB.tell()` will accept.

### Helper function

`atom(prefix, r, c, v)`, where prefix can be either `Is` or `Not`, is a naming helper so propositional symbols need not be typed. It is provided for you in `sudoku_solver.py`; do not change it.

In code, $\mathit{Is}_{rcv}$ and $\mathit{Not}_{rcv}$ are written as single-word symbol names, e.g. `Is3_2_4` and `Not3_2_4`: the prefix is followed immediately by `r`, then by `c` and `v` separated by underscores. No separator is needed between the prefix and `r` since `expr()` only requires a symbol name to start with an uppercase letter. So `Is3_2_4` is parsed as one valid symbol.


In [3]:
# atom() is provided in sudoku_solver.py and imported above.


## Part A: Design the Sudoku Solver

### A.1) Knowledge Representation - Build KB

Every well-posed Sudoku puzzle satisfies exactly these conditions:

- Each cell is assigned **at least one** value from $\{1, \dots, n\}$.
- Each cell is assigned **at most one** value from $\{1, \dots, n\}$, i.e., it cannot hold two different values at once.
- No two cells in the same row hold the same value.
- No two cells in the same column hold the same value.
- No two cells in the same box hold the same value.
- The **givens** cells hold their stated values.

Formalize these Sudoku constraints as propositional logic, in **two** representations:

**(a) General clauses -  `build_general_kb`.**

Encode each of the following directly: no restriction here; you may use arbitrary disjunctions of positive or negated literals. Must return a `PropKB`. 

**(b) Definite (Horn) clauses:- `build_definite_kb`.** Must return a `PropDefiniteKB`. Recall a definite clause is a disjunction of literals with exactly one *positive* literal.

Equivalently written as an implication whose conclusion is a single positive literal and whose premises are a conjunction of positive literals: `P1 & P2 & ... & Pk ==> Q`.

`PropDefiniteKB.tell()` will reject anything else.

Both functions take the puzzle's givens as fixed facts.

- **Implement `build_general_kb()` and `build_definite_kb()` in sudoku_solver.py**
- **Rerun the import cell near the top of this notebook after making changes.**
- **Explain your representation in Conceptual Question 1 below.**

In [4]:
# Implement build_general_kb() and build_definite_kb() in sudoku_solver.py.
# Rerun the import cell near the top of this notebook after making changes.


### A.2) Solve the Puzzle

**(a) Resolution and model checking on the general KB.**

Each unmodified library algorithm runs in a separate process with a five-second wall-time budget. A watchdog inside the child samples its peak resident memory and exits above 256 MiB. Sampling can overshoot this threshold. A resource-budget stop is inconclusive, not a False entailment verdict.


In [5]:
import subprocess
import sys
import textwrap

def run_general_experiment(method, seconds=5.0, memory_mib=256):
    child_code = textwrap.dedent(f"""
        import json, os, sys, threading, time, resource
        def watch_memory():
            while True:
                rss = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
                mib = rss / (1024 ** 2 if sys.platform == 'darwin' else 1024)
                if mib > {memory_mib}:
                    print('Stopped at sampled memory budget; peak RSS=' + str(round(mib, 1)) + ' MiB', flush=True)
                    os._exit(0)
                time.sleep(0.01)
        threading.Thread(target=watch_memory, daemon=True).start()
        from sudoku_solver import *
        data = json.load(open('puzzles.json'))
        givens = {{tuple(map(int, k.split('_'))): v
                  for k, v in data['puzzles'][0]['givens'].items()}}
        kb = build_general_kb(data['n'], data['box_h'], data['box_w'], givens)
        query = atom('Is', 1, 1, 1)
        result = ({method}(kb, query) if '{method}' == 'pl_resolution'
                  else tt_entails(associate('&', kb.clauses), query))
        print('Returned ' + str(result), flush=True)
    """)
    started = time.perf_counter()
    process = subprocess.Popen([sys.executable, '-c', child_code],
                               stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    try:
        stdout, stderr = process.communicate(timeout=seconds)
        if process.returncode:
            raise RuntimeError(stderr.decode())
        status = stdout.decode().strip()
    except subprocess.TimeoutExpired:
        process.kill()
        process.communicate()
        status = 'Stopped at wall-time budget'
    finally:
        if process.poll() is None:
            process.kill()
            process.wait()
    elapsed = time.perf_counter() - started
    print(f'{method}: {status}; elapsed={elapsed:.3f}s')
    return {'method': method, 'status': status, 'seconds': elapsed}

general_experiments = [run_general_experiment(method)
                       for method in ('pl_resolution', 'tt_entails')]


pl_resolution: Stopped at sampled memory budget; peak RSS=288.1 MiB; elapsed=0.584s
tt_entails: Stopped at wall-time budget; elapsed=5.004s


### Observation notes

Record what you observed from the resolution/model-checking experiment here. Use these observations when answering Conceptual Question 2 below.


Both algorithms were actually invoked on `Is1_1_1` for Puzzle 1. The output immediately above records their measured outcomes. A stopped run is inconclusive about entailment; it demonstrates that this library implementation did not return within the stated resource budget. These experiments do not establish a general runtime lower bound.

Recorded outcomes:

```text
pl_resolution: Stopped at sampled memory budget; peak RSS=288.1 MiB; elapsed=0.584s
tt_entails: Stopped at wall-time budget; elapsed=5.004s
```


**(b) Forward chaining on the full grid --** Implement `solve_full_grid_fc()` in sudoku_solver.py. Using your above `build_definite_kb` and the library's `pl_fc_entails` (no need to reimplement them), solve the whole puzzle. Find the value that's entailed for every cell. Reconstruct and display the solved grid.

**(c) Backward chaining -- implement it yourself.**
```python
pl_bc_entails(kb, query) -> bool
```
Implement `pl_bc_entails()` in sudoku_solver.py. Start from the query and recursively try to prove each premise of a rule whose conclusion matches the current goal, bottoming out at known facts. Your function must agree with `pl_fc_entails` on every cell/value pair in this puzzle including correctly returning `False` for values that are *not* part of the solution.

**(d) Backward chaining on the full grid --** Implement `solve_full_grid_bc` in sudoku_solver.py. Using your above `build_definite_kb` and your own `pl_bc_entails`, solve the whole puzzle the same way `solve_full_grid_fc` does: for every cell, try each candidate value until `pl_bc_entails` confirms one. Time both `solve_full_grid_fc` and `solve_full_grid_bc` on the same puzzle and compare. Use your measured result where relevant in Conceptual Question 5.


In [6]:
# Implement solve_full_grid_fc(), pl_bc_entails(), and solve_full_grid_bc()
# in sudoku_solver.py. Rerun the import cell near the top after making changes.


### Verifying the algorithms

Uncomment the following block of code to validate your code.

In [7]:
started = time.perf_counter()
solved = solve_full_grid_fc(n, box_h, box_w, givens)
fc_time = time.perf_counter() - started
assert solved == puzzle['solution']
started = time.perf_counter()
solved_bc = solve_full_grid_bc(n, box_h, box_w, givens)
bc_time = time.perf_counter() - started
assert solved_bc == puzzle['solution']
print(f"solve_full_grid_fc: {fc_time:.4f}s")
print(f"solve_full_grid_bc: {bc_time:.4f}s")
print('Solved grid (FC and BC agree):')
for r in range(1, n + 1):
    print(' '.join(str(solved[(r, c)]) for c in range(1, n + 1)))

solve_full_grid_fc: 3.6598s
solve_full_grid_bc: 0.5474s
Solved grid (FC and BC agree):
1 3 6 9 8 5 4 7 2
2 9 4 3 1 7 8 6 5
5 8 7 2 6 4 1 3 9
7 5 1 4 9 6 2 8 3
3 6 9 8 7 2 5 1 4
4 2 8 5 3 1 7 9 6
8 1 2 6 4 9 3 5 7
6 4 3 7 5 8 9 2 1
9 7 5 1 2 3 6 4 8


## Part B: Conceptual Questions

Answer all five questions directly in this notebook. Replace each **Your answer:** placeholder with your own response.


### 1. Detailed Representation Strategy: General vs. Definite (Horn) Encoding

Explain in detail how you formalized the Sudoku puzzle constraints into propositional logic across both Knowledge Base representations:

**(a) General KB Strategy (`build_general_kb`):** Detail how standard Sudoku rules (e.g., at-least-one value per cell, at-most-one value per cell, row/column/box uniqueness) are directly translated into Conjunctive Normal Form (CNF) clauses without structural restrictions.

**(b) Definite KB Strategy (`build_definite_kb`):** Definite/Horn clauses strictly permit at most one positive literal per clause, prohibiting disjunctive constraints like $(Is_{r,c,1} \lor Is_{r,c,2} \lor Is_{r,c,3} \lor ... \lor Is_{r,c,n})$. Explain step-by-step how your encoding deals with this issue.


## 1. Detailed representation strategy

### General KB (`build_general_kb`)

The general representation uses only the 729 symbols $Is_{r,c,v}$. Every Sudoku condition is placed directly in CNF:

1. **Given:** $Is_{r,c,v}$.
2. **At least one value per cell:** $Is_{r,c,1}\lor\cdots\lor Is_{r,c,9}$.
3. **At most one value per cell:** $\neg Is_{r,c,v_1}\lor\neg Is_{r,c,v_2}$ for every $v_1<v_2$.
4. **Row uniqueness:** $\neg Is_{r,c_1,v}\lor\neg Is_{r,c_2,v}$ for every $c_1<c_2$.
5. **Column uniqueness:** $\neg Is_{r_1,c,v}\lor\neg Is_{r_2,c,v}$ for every $r_1<r_2$.
6. **Box uniqueness:** the analogous binary conflict clause for distinct cells in the same box. Pairs already covered by a row or column clause are skipped to avoid duplicates.

For a 9×9 puzzle this gives 81 at-least-one clauses, 2,916 at-most-one clauses, 2,916 row clauses, 2,916 column clauses, and 1,458 box-only clauses: **10,287 fixed clauses plus the givens**.

### Definite/Horn KB (`build_definite_kb`)

A definite clause cannot express the positive disjunction “one of these nine values holds.” We therefore use two *positive symbol families*: $Is_{r,c,v}$ and $Not_{r,c,v}$. `Not` is part of an atom name, not negation-as-failure.

1. Each given is an `Is` fact.
2. $Is_{r,c,v}\Rightarrow Not_{r,c,v'}$ eliminates other values in the same cell.
3. $Is_{r,c,v}\Rightarrow Not_{r,c',v}$, and the corresponding column and box rules, eliminate the same value from peers.
4. $(\bigwedge_{v'\ne v}Not_{r,c,v'})\Rightarrow Is_{r,c,v}$ implements last-candidate reasoning.

The fixed rule counts are 5,832 same-cell eliminations, 5,832 row eliminations, 5,832 column eliminations, 2,916 box-only eliminations, and 729 last-candidate rules: **21,141 fixed clauses plus the givens**.


This definite KB represents the permitted elimination strategy rather than a logically equivalent translation of every general CNF model. `Not` atoms are explicit positive facts, not built-in Boolean negations. Last-candidate rules do not enforce an unconditional positive disjunction; a puzzle can remain partially unsolved when no rule fires. On consistent givens, deductions are sound, and all supplied puzzles are verified below.


### 2. Theoretical Completeness vs. Computational Tractability

Model checking and resolution-refutation are sound and complete—they are guaranteed to terminate with a correct answer for any propositional KB. Despite this guarantee, explain whether you would use either as the default algorithm for solving Sudoku puzzles. *(Hint: Consider space/time complexity and state-space growth, and use your observations from the experiment above where relevant.)*


## 2. Theoretical completeness versus computational tractability

I would not use these truth-table or resolution implementations as the default full-grid solver. The general KB contains 729 `Is` symbols, so truth-table enumeration has up to $2^{729}$ assignments, with formula evaluation at each model. Its completeness does not make that search practical.

Puzzle 1 starts with 10,317 CNF clauses. `pl_resolution` adds the negated query and explicitly constructs every clause pair in each round: the initial 10,318 clauses alone create 53,225,403 pairs. New resolvents further increase time and memory; the supplied implementation also performs list-based duplicate checks. Worst-case propositional resolution is exponential.

The executed experiment above reports actual elapsed time and resource-budget stops separately for the two algorithms. Such stops are inconclusive about the query, but explain why these implementations are unsuitable as default interactive solvers. Horn chaining is practical for this restricted elimination encoding; completeness for the Horn KB does not mean the encoding solves every valid Sudoku.

## Solve and validate every supplied puzzle

Both solvers build a fresh definite KB and query candidates in row/column/value order. FC calls the supplied `pl_fc_entails`; a KB subclass indexes `clauses_with_premise` without changing the library inference algorithm. BC recursively proves matching rules and shares positive proof results between queries on the same unchanged KB. Neither solver reads the reference solution; it is used only by the assertions below.

In [8]:
import statistics

timing_rows = []
for index, item in enumerate(puzzle_pool, start=1):
    started = time.perf_counter()
    solved_fc = solve_full_grid_fc(n, box_h, box_w, item['givens'])
    fc_time = time.perf_counter() - started

    started = time.perf_counter()
    solved_bc = solve_full_grid_bc(n, box_h, box_w, item['givens'])
    bc_time = time.perf_counter() - started

    assert solved_fc == item['solution']
    assert solved_bc == item['solution']
    timing_rows.append((index, item['given_count'], fc_time, bc_time))
    print(f'Puzzle {index}: givens={item["given_count"]}, FC={fc_time:.4f}s, BC={bc_time:.4f}s, both correct')

print(f'Median FC: {statistics.median(row[2] for row in timing_rows):.4f}s')
print(f'Median BC: {statistics.median(row[3] for row in timing_rows):.4f}s')


Puzzle 1: givens=30, FC=3.8027s, BC=0.5467s, both correct
Puzzle 2: givens=33, FC=3.6837s, BC=0.5339s, both correct
Puzzle 3: givens=36, FC=3.3159s, BC=0.5146s, both correct
Puzzle 4: givens=39, FC=2.9716s, BC=0.4954s, both correct
Puzzle 5: givens=42, FC=2.7336s, BC=0.4275s, both correct
Median FC: 3.3159s
Median BC: 0.5146s


In [9]:
# Every candidate, including False queries, is checked against library FC.
# Repeat for every supplied puzzle; the reference solution is used only to test.
checked = 0
for index, item in enumerate(puzzle_pool, 1):
    definite_kb = build_definite_kb(n, box_h, box_w, item['givens'])
    for (r, c), correct_value in item['solution'].items():
        for candidate in range(1, n + 1):
            query = atom('Is', r, c, candidate)
            backward = pl_bc_entails(definite_kb, query)
            forward = pl_fc_entails(definite_kb, query)
            assert backward == forward == (candidate == correct_value)
            checked += 1
    print(f'Puzzle {index}: all {n ** 3} FC/BC candidate queries agree with the reference.')
print(f'Passed all {checked} true/false cell-value comparisons.')

Puzzle 1: all 729 FC/BC candidate queries agree with the reference.
Puzzle 2: all 729 FC/BC candidate queries agree with the reference.
Puzzle 3: all 729 FC/BC candidate queries agree with the reference.
Puzzle 4: all 729 FC/BC candidate queries agree with the reference.
Puzzle 5: all 729 FC/BC candidate queries agree with the reference.
Passed all 3645 true/false cell-value comparisons.


### 3. Backward Chaining: Design, Pseudocode, and Challenges

Write pseudocode for `pl_bc_entails(kb, query)`, the backward-chaining algorithm you implemented. Show, at a level of detail that reveals the algorithm's structure (not full Python), how the function checks whether the query is already a known fact, finds candidate rules whose conclusion matches the current goal, recursively proves each premise of such a rule, and combines results—both across the premises of one rule and across multiple candidate rules—to reach a single boolean answer.

Then, in your own words, discuss the design challenges you had to work through to make your algorithm both correct and guaranteed to terminate on every puzzle, and explain how your pseudocode addresses them.


## 3. Backward chaining: design, pseudocode, and challenges

```text
BC-ENTAILS(KB, query):
    if KB clauses changed: rebuild head index; proved = facts; clear cached false queries
    if query in proved: return true
    if query is a cached false query: return false
    repeat:
        previous_count = size(proved)
        if PROVE(query, empty visiting set, empty failed set): return true
        if size(proved) == previous_count:
            cache query as false
            return false

PROVE(goal, visiting, failed):
    if goal in proved: return true
    if goal in visiting or failed: return false
    add goal to visiting
    for each rule with conclusion goal:                  // OR across rules
        if PROVE(p, visiting, failed) for every premise p: // AND within rule
            remove goal from visiting
            add goal to proved
            return true
    remove goal from visiting
    add goal to failed
    return false
```

Rules are indexed by conclusion. Proof search starts from the requested goal and recursively reaches facts; there is no forward-closure prepass. Positive results are safe to share because a successful rule has all its premises proved.

Cycles require special care. Encountering a goal on the current proof path rejects only that cyclic branch. A failure table avoids repeating failed searches within one pass, but a failed branch is not permanently false: another branch may establish a missing premise later. Whenever the pass proves new atoms, the query is retried with a fresh failure table. Only an unsuccessful pass with no new proved atoms establishes a negative result. Cache entries are discarded if the clause sequence changes.

For a finite KB, each recursive path contains no repeated goal, each pass explores finitely many rules, and each continuing pass must add a new proved atom. Thus the procedure terminates mathematically (subject to the host language's recursion limit for unusually deep KBs). This retry strategy preserves alternatives that naive global failure memoization would incorrectly discard.

### 4. Expressive Limits of Horn Logic

Named elimination techniques such as Naked Pairs and X-Wing can, in fact, be encoded as definite clauses, using the same `Is`/`Not` vocabulary as your `build_definite_kb`. Work out how you would encode one of these techniques as definite clauses, and discuss the consequences of doing so.


## 4. Expressive limits of Horn logic: Naked Pairs

Consider two cells $a$ and $b$ in one row, column, or box, and two candidate values $x,y$. If every other value has been eliminated from both cells, they form a Naked Pair. For every other cell $d$ in that unit, use two definite clauses:

$$
\left(\bigwedge_{z\notin\{x,y\}}Not_{a,z}\right)\land
\left(\bigwedge_{z\notin\{x,y\}}Not_{b,z}\right)
\Rightarrow Not_{d,x}
$$

$$
\left(\bigwedge_{z\notin\{x,y\}}Not_{a,z}\right)\land
\left(\bigwedge_{z\notin\{x,y\}}Not_{b,z}\right)
\Rightarrow Not_{d,y}
$$

These are valid definite clauses because each `Not` expression is a positive propositional atom and each rule has one positive conclusion. The limitation is size rather than expressibility. Directly expanding the schema over 27 units, 36 cell pairs, 36 value pairs, seven remaining cells, and two conclusions creates $27\times36\times36\times7\times2=489{,}888$ rules before removing overlap. This substantially increases construction time, memory use, indexing cost, and the length of explanations. A production solver would generate or index such rules lazily.


### 5. Data-Driven vs. Goal-Driven Performance

Forward chaining (data-driven) and backward chaining (goal-driven) are both sound and complete for Horn KBs, but their execution runtimes vary depending on the target query.

**(a)** Describe a scenario—in terms of total KB size versus the query-relevant subset—where backward chaining is significantly faster than forward chaining.

**(b)** Describe a scenario where backward chaining offers no performance advantage, or performs worse than forward chaining.

Use your measured forward- vs. backward-chaining result where relevant.


## 5. Data-driven versus goal-driven performance

**(a)** BC can be faster when a large KB has many unrelated components and the query depends on a small proof tree. With an existing head index, BC visits that tree instead of propagating all facts. First-call index construction still scans the KB and should not be ignored in timing.

**(b)** BC may offer no advantage when the query touches most rules, when cycles cause retries, or when many queries repeat the same dependencies. A single FC saturation can serve many queries efficiently. Our required library FC interface instead starts a fresh inference pass for each candidate; our BC implementation reuses proved atoms across queries on an unchanged KB. Both solvers include KB construction in their full-grid timings.

Use the measured table in “Solve and validate every supplied puzzle” above: it runs both methods on identical inputs and checks identical solutions. The comparison concerns these implementations, including the premise index for FC and proof cache for BC, and does not establish a universal ordering of the algorithms.

In this recorded run, the median over five puzzles was **FC 3.3159 s** and **BC 0.5146 s** (FC/BC ratio approximately 6.44). BC was faster in this implementation because its proof cache is reused, whereas the supplied FC function restarts for each candidate. This does not contradict the advantage of a single shared FC saturation in a different implementation.


## Part C: Streamlit Integration

Wrap your Sudoku solver and inference logic into an interactive Streamlit application, `sudoku_app.py`. Your application must implement the following:

**1. Puzzle selection & visual board display**
- An interactive selector/dropdown to pick any puzzle from `puzzles.json`.
- A visual rendering of the grid that clearly distinguishes the initial *givens* from empty cells.

**2. Full-grid auto-solver, with algorithm selection**
- A control (e.g. radio buttons) letting the user choose forward chaining (`solve_full_grid_fc`) or backward chaining (`solve_full_grid_bc`) before solving.
- A button that solves the full grid with the chosen algorithm, renders the solved state, and displays how long the solve took -- so the timing gap from task (d) is visible in the app, not just in the notebook.

**3. Targeted cell entailment query**
- Inputs for row ($r$), column ($c$), and value ($v$).
- A button that checks whether $\mathit{Is}_{rcv}$ is entailed (using `pl_bc_entails`) and displays the boolean verdict (`True` / `False`).

**4. Reasoning trace ("tutor mode")**
- Instrument your chosen inference algorithm (forward or backward chaining) to record the reasoning steps it takes while answering a query.
- Present that trace in a human-readable format -- not a raw Python string or internal symbol dictionary. For example: expandable cards/accordions showing the rule-firing sequence (*"Inferred $\mathit{Not}_{1,2,3}$ because row 1 already contains value 3"* $\implies$ *"Deduce $\mathit{Is}_{1,2,4}$ as the last remaining candidate"*), or plain-English sentences explaining each elimination by row, column, or box constraint.

Import `atom`, `build_definite_kb`, `build_general_kb`, `solve_full_grid_fc`, `solve_full_grid_bc`, and `pl_bc_entails` from `sudoku_solver.py`. Do not copy or rewrite these core functions in the Streamlit file. You may add app-specific helper functions where needed for the interface or reasoning trace.

Refer to the provided `StreamlitDeploymentGuide.pdf` guide to deploy your code and include the link for your Streamlit app below.


## Submission

Submit only the following three files:

1. `Sudoku_Assignment.ipynb` — with all required cells run, outputs visible, and all conceptual questions answered.
2. `sudoku_solver.py` — containing your implementation of the knowledge-base and inference functions.
3. `sudoku_app.py` — containing your Streamlit application.

The provided support files are required to run the assignment but are not part of the student submission.

### Deployed Streamlit app URL

Paste your deployed Streamlit app URL here:

`https://sudoku-ai-solver.streamlit.app/`
